content/articles/automation_lessons.md

```diff
@@ -0,0 +1,110 @@
+
+Even when it doesn't turn out to be a lie, eventually you'll need test and
+installation automation. Paying the price up front has many advantages:
+
+-   There is a significant risk that there won't be time later.
+
+-   Existing test and installation automation provides an exceptionally valuable
+    design constraint when adding or modifying features. If something is hard to
+    test or hard to install (or upgrade) as designed then that design is _bad_.
+    Existing harnesses help you recognize when that is about to happen.
+
+-   The time investment is amortized against the savings gained, sooner.
+
+With regards to "too complicated to automate". Sit with that for a moment. If
+that's true, what does it say a) about trying to use the software, b) trying to
+maintain the software over the long run. Let's be honest here.
+
+Okay, say you've managed time well enough to have built tooling that can
+automatically test the software to a significant degree as well as
+automatically install it somewhere. Great. Then that tooling should be required
+to run (automatically) for any pull request and required to pass all tests and
+properly do an install before that pull request can merge to its target.
+Ideally it would also do an upgrade. That is what _continuous_ integration
+means.
+
+We've argued about this being too human-time and compute-resource consuming
+many times over. Meanwhile:
+
+-   We encourage developers to run cloud VMs for their own development and manual
+    testing work at a vast aggregate monetary cost. Effective local (to the
+    laptop!) testing (to be covered in future writing) combined with
+    _comprehensive_ centralized testing would be much cheaper.
+
+-   Most developers are probably working multiple independent tasks. Going into a
+    wait state for testing to complete and not having something else to do in the
+    meantime is unlikely. If it is happening then it suggests that tasks are not
+    being decomposed and planned effectively.
+
+-   If the waiting is because tests or testing infrastructure is too flakey, or
+    because installation takes too long, imagine what that implies for the
+    experience of the customer. Flakey tests and slow installation are problems
+    to be solved, not worked around, because if those problems exist for you,
+    then they will exist for other people, people who pay.
+
+-   You, individually, spending some time waiting if it means other people do not
+    have to wait, later, and do not have to experience bugs, later, is a more
+    than worthwhile trade.
+
+We often find ourselves saying a version of the following:
+
+> Once we have it working in context X we'll look into making it work on pull
+> requests.
+
+If you can only take one thing (but please take more!) from this document take
+the awareness that that is backwards. Instead:
+
+> **Make it work on pull requests. Once it is working there, use it in other
+> contexts.**
```

Comment:
I'm not sure boldification is required here: It might be a design issue, but it seems more distracting than helpful when rendered? Seems to me that just ending with this quote makes the point pretty clearly.